src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs

```diff
@@ -11,18 +11,43 @@ public static class PageDirectiveFeature
     {
         public static bool TryGetRouteTemplate(RazorProjectItem projectItem, out string template)
         {
+            if (projectItem == null)
+            {
+                throw new ArgumentNullException(nameof(projectItem));
+            }
+
             const string PageDirective = "@page";
 
+            var stream = projectItem.Read();
+
+            if (stream == null)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(projectItem)}.{nameof(projectItem.Read)} can't return null.");
+            }
+
             string content;
-            using (var streamReader = new StreamReader(projectItem.Read()))
+            using (var streamReader = new StreamReader(stream))
             {
-                content = streamReader.ReadToEnd();
+                content = streamReader.ReadToEnd().TrimStart();
             }
 
             if (content.StartsWith(PageDirective, StringComparison.Ordinal))
             {
-                var newLineIndex = content.IndexOf(Environment.NewLine, PageDirective.Length);
-                template = content.Substring(PageDirective.Length, newLineIndex - PageDirective.Length).Trim();
+                var endOfDirective = content.IndexOf(Environment.NewLine, PageDirective.Length);
```

Comment:
We need to support all possible newlines on all possible systems. I would suggest creating a `SourceDocument` here and then using `SourceDocument.Lines`. We want to have explicit newline handling in as few places as possible.